File: vespajlib/src/main/java/com/yahoo/slime/Inserter.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.slime;

/**
 * Helper interface for inserting values into any of the container
 * classes (ArrayValue, ObjectValue, or Slime).  May be useful for
 * deserializers where you can use it to decouple the actual value
 * decoding from the container where the value should be inserted.
 */
public interface Inserter {

    Cursor insertNIX();
    Cursor insertBOOL(boolean value);
    Cursor insertLONG(long value);
    Cursor insertDOUBLE(double value);
    Cursor insertSTRING(String value);
    Cursor insertSTRING(byte[] utf8);
    Cursor insertDATA(byte[] value);
    Cursor insertARRAY();
    Cursor insertOBJECT();

}

File: client/go/internal/vespa/slime/inserter.go
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

package slime

type Inserter interface {
	Insert(value Value) Value
}

type myInserter func(value Value) Value

func (f myInserter) Insert(value Value) Value {
	return f(value)
}

func InsertRoot(root *Value) Inserter {
	return myInserter(func(value Value) Value {
		*root = value
		return value
	})
}

func InsertEntry(arr Value) Inserter {
	return myInserter(func(value Value) Value {
		return arr.Add(value)
	})
}

func InsertField(obj Value, name string) Inserter {
	return myInserter(func(value Value) Value {
		return obj.Set(name, value)
	})
}
